Interrupting msgmerge results in an empty .po file

Interrupting execution of `msgmerge` results in an empty .po file:
```
ninja -C build/release/ dtee-update-po
ninja: Entering directory `build/release/'
[1/1] Running external command dtee-update-po.
^CTraceback (most recent call last):
  File "/home/simon/.local/bin/meson", line 11, in <module>
    sys.exit(main())
  File "/home/simon/.local/lib/python3.5/site-packages/mesonbuild/mesonmain.py", line 392, in main
Traceback (most recent call last):
  File "/home/simon/.local/bin/meson", line 11, in <module>
    return run(sys.argv[1:], launcher)
  File "/home/simon/.local/lib/python3.5/site-packages/mesonbuild/mesonmain.py", line 344, in run
    sys.exit(main())
  File "/home/simon/.local/lib/python3.5/site-packages/mesonbuild/mesonmain.py", line 392, in main
    sys.exit(run_script_command(args[1:]))
  File "/home/simon/.local/lib/python3.5/site-packages/mesonbuild/mesonmain.py", line 254, in run_script_command
    return run(sys.argv[1:], launcher)
  File "/home/simon/.local/lib/python3.5/site-packages/mesonbuild/mesonmain.py", line 344, in run
    return cmdfunc(cmdargs)
  File "/home/simon/.local/lib/python3.5/site-packages/mesonbuild/scripts/commandrunner.py", line 70, in run
    pc.wait()
  File "/usr/lib/python3.5/subprocess.py", line 1658, in wait
    sys.exit(run_script_command(args[1:]))
  File "/home/simon/.local/lib/python3.5/site-packages/mesonbuild/mesonmain.py", line 254, in run_script_command
    return cmdfunc(cmdargs)
  File "/home/simon/.local/lib/python3.5/site-packages/mesonbuild/scripts/gettext.py", line 113, in run
    return update_po(src_sub, options.pkgname, langs)
  File "/home/simon/.local/lib/python3.5/site-packages/mesonbuild/scripts/gettext.py", line 74, in update_po
    subprocess.check_call(['msgmerge', '-q', '-o', pofile, pofile, potfile])
  File "/usr/lib/python3.5/subprocess.py", line 576, in check_call
    (pid, sts) = self._try_wait(0)
  File "/usr/lib/python3.5/subprocess.py", line 1608, in _try_wait
    retcode = call(*popenargs, **kwargs)
  File "/usr/lib/python3.5/subprocess.py", line 559, in call
    return p.wait(timeout=timeout)
  File "/usr/lib/python3.5/subprocess.py", line 1658, in wait
    (pid, sts) = os.waitpid(self.pid, wait_flags)
KeyboardInterrupt
    (pid, sts) = self._try_wait(0)
  File "/usr/lib/python3.5/subprocess.py", line 1608, in _try_wait
    (pid, sts) = os.waitpid(self.pid, wait_flags)
KeyboardInterrupt
ninja: build stopped: interrupted by user.
GNUmakefile:98: recipe for target 'po' failed
make: *** [po] Error 2
```

The `fr.po` file is now empty:
```
-rw-rw-r-- 1 simon simon 4698 Dec 28 17:44 en.po
-rw-rw-r-- 1 simon simon    0 Dec 28 17:44 fr.po
```

The output file should be created with a different name and renamed into place so that this doesn't happen.

--- a/docs/markdown/snippets/werror-linker.md
+++ b/docs/markdown/snippets/werror-linker.md
@@ -1,0 +1,6 @@
+## `werror=true` now applies to the linker as well
+
+When `werror=true` is set, Meson now passes the appropriate
+fatal-warnings flag to the linker (for example `--fatal-warnings`
+for GNU ld, `-fatal_warnings` for Apple ld, `/WX` for MSVC link).
+Previously, `werror=true` only affected compiler warnings.

--- a/docs/yaml/functions/_build_target_base.yaml
+++ b/docs/yaml/functions/_build_target_base.yaml
@@ -43,15 +43,9 @@ kwargs:
     description: precompiled header file to use for the given language
 
   <lang>_args:
-    type: array[str]
-    description: |
-      compiler flags to use for the given language;
-      eg: `cpp_args` for C++
-
-  vala_args:
     type: array[str | file]
     description: |
-      Compiler flags for Vala. Unlike other languages this may contain Files
+      compiler flags to use for the given language; eg: `cpp_args` for C++
 
   sources:
     type: str | file | custom_tgt | custom_idx | generated_list | structured_src

--- a/mesonbuild/backend/backends.py
+++ b/mesonbuild/backend/backends.py
@@ -783,7 +783,7 @@ class Backend:
                     rel_src = rel_src[len(build_dir) + 1:]
                 rel_src = os.path.relpath(rel_src, self.get_target_private_dir(target))
             else:
-                rel_src = os.path.basename(rel_src)
+                rel_src = os.path.relpath(rel_src, os.path.join(self.build_to_src, target.get_subdir()))
             # A meson- prefixed directory is reserved; hopefully no-one creates a file name with such a weird prefix.
             gen_source = 'meson-generated_' + rel_src[:-5] + '.c'
         elif source.is_built:

--- a/mesonbuild/backend/vs2010backend.py
+++ b/mesonbuild/backend/vs2010backend.py
@@ -1449,6 +1449,8 @@ class Vs2010Backend(backends.Backend):
         link = ET.SubElement(compiles, 'Link')
         extra_link_args = compiler.compiler_args()
         extra_link_args += compiler.get_optimization_link_args(self.optimization)
+        if self.get_target_option(target, 'werror'):
+            extra_link_args += compiler.get_linker_fatal_warnings()
         # Generate Debug info
         if self.debug:
             self.generate_debug_information(link)

--- a/mesonbuild/cmake/common.py
+++ b/mesonbuild/cmake/common.py
@@ -5,14 +5,17 @@ from __future__ import annotations
 
 from ..mesonlib import MesonException
 from ..options import OptionKey
+from ..tooldetect import detect_ninja
 from .. import mlog
 from pathlib import Path
+import functools
 import typing as T
 
 if T.TYPE_CHECKING:
     from ..compilers.compilers import Language
     from ..environment import Environment
     from ..interpreterbase import TYPE_var
+    from .._typing import ImmutableListProtocol
 
 language_map: T.Mapping[Language, str] = {
     'c': 'C',
@@ -107,11 +110,21 @@ def _flags_to_list(raw: str) -> T.List[str]:
     res = [r for r in res if len(r) > 0]
     return res
 
+@functools.lru_cache(maxsize=None)
+def _cmake_get_generator_args(backend_name: str) -> ImmutableListProtocol[str]:
+    assert backend_name in backend_generator_map
+    args = ['-G', backend_generator_map[backend_name]]
+    if backend_name == 'ninja':
+        ninja = detect_ninja()
+        if ninja:
+            assert len(ninja) == 1
+            args += [f'-DCMAKE_MAKE_PROGRAM={ninja[0]}']
+    return args
+
 def cmake_get_generator_args(env: 'Environment') -> T.List[str]:
     backend_name = env.coredata.optstore.get_value_for(OptionKey('backend'))
     assert isinstance(backend_name, str)
-    assert backend_name in backend_generator_map
-    return ['-G', backend_generator_map[backend_name]]
+    return list(_cmake_get_generator_args(backend_name))
 
 def cmake_defines_to_args(raw: T.List[T.Dict[str, TYPE_var]], permissive: bool = False) -> T.List[str]:
     res: T.List[str] = []

--- a/mesonbuild/compilers/compilers.py
+++ b/mesonbuild/compilers/compilers.py
@@ -356,6 +356,8 @@ def get_base_link_args(target: 'BuildTarget',
                        env: 'Environment') -> T.List[str]:
     args: T.List[str] = []
     build_dir = env.get_build_dir()
+    if env.coredata.get_option_for_target(target, 'werror'):
+        args.extend(linker.get_linker_fatal_warnings())
     try:
         if env.coredata.get_option_for_target(target, 'b_lto'):
             if env.coredata.get_option_for_target(target, 'werror'):
@@ -1179,6 +1181,9 @@ class Compiler(HoldableObject, metaclass=SimpleABC):
 
     def get_optimization_link_args(self, optimization_level: str) -> T.List[str]:
         return self.linker.get_optimization_link_args(optimization_level)
+
+    def get_linker_fatal_warnings(self) -> T.List[str]:
+        return self.linker.fatal_warnings()
 
     def get_soname_args(self, prefix: str, shlib_name: str, suffix: str, soversion: str,
                         darwin_versions: T.Tuple[str, str]) -> T.List[str]:

--- a/mesonbuild/compilers/cuda.py
+++ b/mesonbuild/compilers/cuda.py
@@ -688,6 +688,9 @@ class CudaCompiler(Compiler):
     def get_optimization_link_args(self, optimization_level: str) -> T.List[str]:
         return self._to_host_flags(self.host_compiler.get_optimization_link_args(optimization_level), Phase.LINKER)
 
+    def get_linker_fatal_warnings(self) -> T.List[str]:
+        return self._to_host_flags(self.host_compiler.get_linker_fatal_warnings(), Phase.LINKER)
+
     def build_rpath_args(self, build_dir: str, from_dir: str, target: BuildTarget,
                          extra_paths: T.Optional[T.List[str]] = None
                          ) -> T.Tuple[T.List[str], T.Set[bytes]]:

--- a/mesonbuild/compilers/mixins/islinker.py
+++ b/mesonbuild/compilers/mixins/islinker.py
@@ -111,6 +111,9 @@ class BasicLinkerIsCompilerMixin(Compiler):
     def get_optimization_link_args(self, optimization_level: str) -> T.List[str]:
         return []
 
+    def get_linker_fatal_warnings(self) -> T.List[str]:
+        return []
+
     def get_link_debugfile_name(self, targetfile: str) -> T.Optional[str]:
         return None
 

--- a/mesonbuild/compilers/mixins/visualstudio.py
+++ b/mesonbuild/compilers/mixins/visualstudio.py
@@ -129,6 +129,12 @@ class VisualStudioLikeCompiler(Compiler, metaclass=mesonlib.SimpleABC):
         # TODO: use ImmutableListProtocol[str] here instead
         return self.always_args.copy()
 
+    def get_no_stdinc_args(self) -> T.List[str]:
+        return ['/X']
+
+    def get_no_stdlib_link_args(self) -> T.List[str]:
+        return ['/NODEFAULTLIB']
+
     def get_pch_suffix(self) -> str:
         return 'pch'
 

--- a/mesonbuild/compilers/objcpp.py
+++ b/mesonbuild/compilers/objcpp.py
@@ -34,6 +34,12 @@ class ObjCPPCompiler(CLikeCompiler, Compiler):
                           full_version=full_version, linker=linker)
         CLikeCompiler.__init__(self)
 
+    def get_no_stdinc_args(self) -> T.List[str]:
+        return ['-nostdinc++']
+
+    def get_no_stdlib_link_args(self) -> T.List[str]:
+        return ['-nostdlib++']
+
     def form_compileropt_key(self, basename: str) -> OptionKey:
         if basename == 'std':
             return OptionKey('cpp_std', machine=self.for_machine)

--- a/mesonbuild/compilers/rust.py
+++ b/mesonbuild/compilers/rust.py
@@ -471,6 +471,10 @@ class RustCompiler(Compiler):
         return rustc_link_args(super().headerpad_args())
 
     @functools.lru_cache(maxsize=None)
+    def get_linker_fatal_warnings(self) -> T.List[str]:
+        return rustc_link_args(super().get_linker_fatal_warnings())
+
+    @functools.lru_cache(maxsize=None)
     def get_allow_undefined_link_args(self) -> T.List[str]:
         return rustc_link_args(super().get_allow_undefined_link_args())
 

--- a/mesonbuild/interpreter/interpreter.py
+++ b/mesonbuild/interpreter/interpreter.py
@@ -2710,7 +2710,12 @@ class Interpreter(InterpreterBase, HoldableObject):
             output = outputs[0]
             if depfile:
                 depfile = mesonlib.substitute_values([depfile], values)[0]
-        ofile_rpath = os.path.join(self.subdir, output)
+
+        # Validate build_subdir
+        build_subdir = kwargs['build_subdir']
+        self.validate_build_subdir(build_subdir, output)
+
+        ofile_rpath = os.path.join(self.subdir, build_subdir, output)
         if ofile_rpath in self.configure_file_outputs:
             mesonbuildfile = os.path.join(self.subdir, 'meson.build')
             current_call = f"{mesonbuildfile}:{self.current_node.lineno}"
@@ -2719,12 +2724,8 @@ class Interpreter(InterpreterBase, HoldableObject):
         else:
             self.configure_file_outputs[ofile_rpath] = self.current_node.lineno
 
-        # Validate build_subdir
-        build_subdir = kwargs['build_subdir']
-        self.validate_build_subdir(build_subdir, output)
-
-        (ofile_path, ofile_fname) = os.path.split(os.path.join(self.subdir, build_subdir, output))
-        ofile_abs = os.path.join(self.environment.build_dir, ofile_path, ofile_fname)
+        ofile_path, ofile_fname = os.path.split(ofile_rpath)
+        ofile_abs = os.path.join(self.environment.build_dir, ofile_rpath)
         os.makedirs(os.path.split(ofile_abs)[0], exist_ok=True)
 
         # Perform the appropriate action
@@ -2737,7 +2738,7 @@ class Interpreter(InterpreterBase, HoldableObject):
                         raise InvalidArguments(
                             f'"configuration_data": initial value dictionary key "{k!r}"" must be "str | int | bool", not "{v!r}"')
                 conf = build.ConfigurationData(conf)
-            mlog.log('Configuring', mlog.bold(output), 'using configuration')
+            mlog.log('Configuring', mlog.bold(os.path.join(build_subdir, output)), 'using configuration')
             if len(inputs) > 1:
                 raise InterpreterException('At most one input file can given in configuration mode')
             if inputs:

--- a/mesonbuild/interpreter/interpreter.py
+++ b/mesonbuild/interpreter/interpreter.py
@@ -2124,8 +2124,6 @@ class Interpreter(InterpreterBase, HoldableObject):
                                    f'(there are {len(kwargs["install_tag"])} install_tags, '
                                    f'and {len(kwargs["output"])} outputs)')
 
-        for t in kwargs['output']:
-            self.validate_forbidden_targets(t)
         self._validate_custom_target_outputs(len(inputs) > 1, kwargs['output'], "custom_target")
 
         tg = build.CustomTarget(
@@ -2151,6 +2149,10 @@ class Interpreter(InterpreterBase, HoldableObject):
             install_tag=kwargs['install_tag'],
             backend=self.backend,
             build_subdir=kwargs['build_subdir'])
+
+        subdir = tg.get_builddir()
+        for t in tg.get_outputs():
+            self.validate_forbidden_targets(t, not subdir)
         self.add_target(tg.name, tg)
         return tg
 
@@ -3249,8 +3251,7 @@ class Interpreter(InterpreterBase, HoldableObject):
                                            'string or File-type object')
         return results
 
-    @staticmethod
-    def validate_forbidden_targets(name: str) -> None:
+    def validate_forbidden_targets(self, name: str, in_root: bool) -> None:
         if name.startswith('meson-internal__'):
             raise InvalidArguments("Target names starting with 'meson-internal__' are reserved "
                                    "for Meson's internal use. Please rename.")
@@ -3258,8 +3259,12 @@ class Interpreter(InterpreterBase, HoldableObject):
             raise InvalidArguments("Target names starting with 'meson-' and without a file extension "
                                    "are reserved for Meson's internal use. Please rename.")
         if name in coredata.FORBIDDEN_TARGET_NAMES:
-            raise InvalidArguments(f"Target name '{name}' is reserved for Meson's "
-                                   "internal use. Please rename.")
+            if in_root:
+                raise InvalidArguments(f"Target name '{name}' is reserved for Meson's "
+                                       "internal use. Please rename.")
+            else:
+                FeatureNew.single_use(f"Target name '{name}' reserved in the root build directory, but allowed in subdirectories",
+                                      '1.12.0', self.subproject, location=self.current_node)
 
     def add_target(self, name: str, tobj: build.Target) -> None:
         if self.backend.name == 'none':
@@ -3282,11 +3287,12 @@ class Interpreter(InterpreterBase, HoldableObject):
         build_subdir = tobj.get_build_subdir()
         self.validate_build_subdir(build_subdir, name)
 
-        self.validate_forbidden_targets(name)
+        subdir = tobj.get_builddir()
+        self.validate_forbidden_targets(name, not subdir)
+
         # To permit an executable and a shared library to have the
         # same name, such as "foo.exe" and "libfoo.a".
         idname = tobj.get_id()
-        subdir = tobj.get_builddir()
         namedir = (name, subdir)
 
         if idname in self.build.targets:

--- a/mesonbuild/interpreter/type_checking.py
+++ b/mesonbuild/interpreter/type_checking.py
@@ -569,9 +569,9 @@ _VS_MODULE_DEFS_KW: KwargInfo[T.Optional[T.Union[str, File, CustomTarget, Custom
     since_values={CustomTargetIndex: '1.3.0'}
 )
 
-_BASE_LANG_KW: KwargInfo[T.List[str]] = KwargInfo(
+_BASE_LANG_KW: KwargInfo[T.List[T.Union[str, File]]] = KwargInfo(
     'UNKNOWN',
-    ContainerTypeInfo(list, (str)),
+    ContainerTypeInfo(list, (str, File)),
     listify=True,
     default=[],
 )

--- a/mesonbuild/modules/cmake.py
+++ b/mesonbuild/modules/cmake.py
@@ -12,6 +12,7 @@ from . import ExtensionModule, ModuleReturnValue, ModuleObject, ModuleInfo
 from .. import build, mesonlib, mlog, dependencies
 from ..options import OptionKey
 from ..cmake import TargetOptions, cmake_defines_to_args
+from ..dependencies.cmake import CMakeDependency
 from ..interpreter import SubprojectHolder
 from ..interpreter.type_checking import REQUIRED_KW, INSTALL_DIR_KW, INCLUDE_TYPE, NoneType, in_set_validator
 from ..interpreterbase import (
@@ -272,8 +273,16 @@ class CmakeModule(ExtensionModule):
         if not cmakebin.found():
             return False
 
-        p, stdout, stderr = mesonlib.Popen_safe(cmakebin.get_command() + ['--system-information', '-G', 'Ninja'])[0:3]
-        if p.returncode != 0:
+        # Try different CMake generators since specifying no generator may fail
+        # in cygwin for some reason
+        for gen in CMakeDependency.class_cmake_generators:
+            cmd = cmakebin.get_command() + ['--system-information']
+            if gen:
+                cmd += ['-G', gen]
+            p, stdout, stderr = mesonlib.Popen_safe(cmd)[0:3]
+            if p.returncode == 0:
+                break
+        else:
             mlog.log(f'error retrieving cmake information: returnCode={p.returncode} stdout={stdout} stderr={stderr}')
             return False
 

--- a/mesonbuild/modules/gnome.py
+++ b/mesonbuild/modules/gnome.py
@@ -1393,11 +1393,10 @@ class GnomeModule(ExtensionModule):
                                         mesonlib.FileMode(), state.subproject, install_tag='doc')
                 targets.append(l_data)
 
-            po_fname = l + '.po'
-            po_args: CommandList = [
-                msgmerge, '-q', '-o',
-                os.path.join('@SOURCE_ROOT@', l_subdir, po_fname),
-                os.path.join('@SOURCE_ROOT@', l_subdir, po_fname), pot_file]
+            po_path = os.path.join('@SOURCE_ROOT@', l_subdir, l + '.po')
+            po_args: CommandList = [*state.environment.get_build_command(),
+                                    '--internal', 'exe', '--capture', po_path, '--',
+                                    msgmerge, '-q', po_path, pot_file]
             potarget = build.RunTarget(f'help-{project_id}-{l}-update-po',
                                        po_args, [pottarget], l_subdir, state.subproject,
                                        state.environment)

--- a/mesonbuild/mtest.py
+++ b/mesonbuild/mtest.py
@@ -1149,7 +1149,7 @@ class TestRunTAP(TestRun):
         return True
 
     def complete(self) -> None:
-        if self.returncode != 0 and not self.res.was_killed():
+        if self.returncode != 0 and self.res.is_ok():
             self.res = TestResult.ERROR
             self.stde = self.stde or ''
             self.stde += f'\n(test program exited with status code {self.returncode})'

--- a/mesonbuild/mtest.py
+++ b/mesonbuild/mtest.py
@@ -738,7 +738,7 @@ class ConsoleLogger(TestLogger):
             else:
                 print(harness.format(result, mlog.colorize_console(), max_left_width=self.max_left_width),
                       flush=True)
-                if result.verbose or result.res.is_bad():
+                if result.verbose or harness.is_bad_result(result):
                     self.print_log(harness, result)
             if result.warnings:
                 print(flush=True)
@@ -1673,6 +1673,7 @@ class TestHarness:
     def __init__(self, options: argparse.Namespace):
         self.options = options
         self.collected_failures: T.List[TestRun] = []
+        self.maxfail_reached = False
         self.fail_count = 0
         self.expectedfail_count = 0
         self.unexpectedpass_count = 0
@@ -1835,10 +1836,13 @@ class TestHarness:
         else:
             sys.exit(f'Unknown test result encountered: {result.res}')
 
-        if result.res.is_bad():
+        if self.is_bad_result(result):
             self.collected_failures.append(result)
         for l in self.loggers:
             l.log(self, result)
+
+    def is_bad_result(self, result: TestRun) -> bool:
+        return result.res.is_bad() and not (result.res is TestResult.INTERRUPT and self.maxfail_reached)
 
     @property
     def numlen(self) -> int:
@@ -2154,6 +2158,7 @@ class TestHarness:
                 self.process_test_result(res)
                 maxfail = self.options.maxfail
                 if maxfail and self.fail_count >= maxfail and res.res.is_bad():
+                    self.maxfail_reached = True
                     cancel_all_tests()
 
         def test_done(f: asyncio.Future) -> None:

--- a/mesonbuild/scripts/gettext.py
+++ b/mesonbuild/scripts/gettext.py
@@ -8,6 +8,9 @@ import argparse
 import subprocess
 import typing as T
 
+from .meson_exe import run_exe
+from ..utils.core import ExecutableSerialisation
+
 parser = argparse.ArgumentParser()
 parser.add_argument('command')
 parser.add_argument('--pkgname', default='')
@@ -59,7 +62,9 @@ def update_po(src_sub: str, msgmerge: str, msginit: str, pkgname: str, langs: T.
     for l in langs:
         pofile = os.path.join(src_sub, l + '.po')
         if os.path.exists(pofile):
-            subprocess.check_call([msgmerge, '-q', '-o', pofile, pofile, potfile])
+            exe = ExecutableSerialisation([msgmerge, '-q', pofile, potfile], capture=pofile)
+            if run_exe(exe) != 0:
+                return 1
         else:
             subprocess.check_call([msginit, '--input', potfile, '--output-file', pofile, '--locale', l, '--no-translator'])
     return 0

--- a/mesonbuild/tooldetect.py
+++ b/mesonbuild/tooldetect.py
@@ -202,9 +202,9 @@ def detect_scanbuild() -> T.List[str]:
                 break
 
     if exelist:
-        tool = exelist[0]
-        if os.path.isfile(tool) and os.access(tool, os.X_OK):
-            return [tool]
+        tool = shutil.which(exelist[0])
+        if tool is not None:
+            return [tool] + exelist[1:]
     return []
 
 def detect_clangformat() -> T.List[str]:

--- a/test cases/vala/33 subdir sources and both_libraries/lib/meson.build
+++ b/test cases/vala/33 subdir sources and both_libraries/lib/meson.build
@@ -1,0 +1,10 @@
+args = []
+# https://github.com/mesonbuild/meson/issues/1969
+if get_option('unity') == 'on'
+  vala_args = ['-H', 'mylib.h']
+endif
+
+l = both_libraries('valalib', 'subdir/mylib.vala',
+                   vala_args : args,
+                   dependencies : valadeps,
+                   install_dir : [false, get_option('includedir'), false])

--- a/test cases/vala/33 subdir sources and both_libraries/lib/subdir/mylib.vala
+++ b/test cases/vala/33 subdir sources and both_libraries/lib/subdir/mylib.vala
@@ -1,0 +1,5 @@
+public class LibraryObject : Object {
+    public void func() {
+       stdout.printf("Method in library called.");
+    }
+}

--- a/test cases/vala/33 subdir sources and both_libraries/meson.build
+++ b/test cases/vala/33 subdir sources and both_libraries/meson.build
@@ -1,0 +1,5 @@
+project('shared library', 'vala', 'c')
+
+valadeps = [dependency('glib-2.0'), dependency('gobject-2.0')]
+
+subdir('lib')

